### DISCUSSION TITLE
Splitting fe.[ch]pp

### DISCUSSION THREAD

**Discussion by mlstowell** [mfem/mfem#2322] [Ideas]
There seems to be a consensus forming around the idea that `fe.cpp` and `fe.hpp` are growing too large.
```
[pform:GitHub/mfem-master/fem] stowell1% wc -l fe.?pp
   13489 fe.cpp
    3493 fe.hpp
   16982 total
```

There are many ways these might be split into more manageable sizes but to get the conversation started I'll propose the following:
- fe.[ch]pp to contain the FiniteElement base class and any other common definitions or code
- fe_h1.[ch]pp to contain the H1 basis functions
- fe_l2.[ch]pp to contain the L2 basis functions
- fe_nd.[ch]pp to contain the ND basis functions
- fe_rt.[ch]pp to contain the RT basis functions
- fe_misc.[ch]pp to contain other classes that don't fit naturally in the above files

Other ideas are certainly welcome. For something like this there seems to be little point in starting a branch or a PR until we reach some sort of an agreement on how best to split things. I look forward to hearing people's ideas on this. 

Attn: @dylan-copeland, @tzanio, @v-dobrev

**Comment by v-dobrev** [comment]
This sounds good to me. The only remaining question for me is what classes and functions stay in `fe.[ch]pp` and what go to `fe_misc.[ch]pp`. If there are just a few things to go in `fe_misc.[ch]pp` we can just keep them in `fe.[ch]pp`.

**Comment by tzanio** [comment]
Mark, can you take a crack at your proposed split when you get a chance and report the resulting sizes?

My feeling is that `fe_misc.[ch]pp` will probably be the largest.

**Reply by mlstowell** [reply]
A draft of this suggestion (with a few modifications) has been started in PR #2335
